src/DataStructures/Tensor/EagerMath/Norms.hpp

```diff
@@ -0,0 +1,150 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <string>
+
+#include "DataStructures/DataBox/DataBoxTag.hpp"
+#include "DataStructures/DataVector.hpp"
+#include "DataStructures/Tensor/Tensor.hpp"
+#include "Utilities/Functional.hpp"
+#include "Utilities/Numeric.hpp"
+#include "Utilities/TMPL.hpp"
+
+namespace L2Norm_detail {
+template <typename DataType, typename Symm, typename IndexList>
+Scalar<DataType> pointwise_l2_norm_square(
+    const Tensor<DataType, Symm, IndexList>& tensor) noexcept {
+  auto pointwise_l2_normsq = make_with_value<Scalar<DataType>>(tensor, 0.);
+  for (auto tensor_element = tensor.begin(); tensor_element != tensor.end();
+       ++tensor_element) {
+    // In order to handle tensory symmetries, we multiply the square of each
+    // stored component with its multiplicity
+    get(pointwise_l2_normsq) +=
+        tensor.multiplicity(tensor_element) * square(*tensor_element);
+  }
+  return pointwise_l2_normsq;
+}
+}  // namespace L2Norm_detail
+
+/*!
+ * \ingroup TensorGroup
+ * \brief Compute point-wise Euclidean \f$L^2\f$-norm of arbitrary Tensors.
+ *
+ * \details
+ * At each grid point \f$p\f$ in the element, this function computes the
+ * point-wise Frobenius norm of a given Tensor with arbitrary rank. If the
+ * Tensor \f$A\f$ has rank \f$n\f$ and dimensionality \f$D\f$, then its
+ * Frobenius norm at point \f$p\f$ is computed as:
+ *
+ * \f{equation}
+ * ||A||_2(p) =
+ *    \left(\sum^{D-1}_{i_1=0}\sum^{D-1}_{i_2=0}\cdots \sum^{D-1}_{i_n=0}
+ *          |A_{i_1 i_2 \cdots i_n}(p)|^2 \right)^{1/2},
+ * \f}
+ *
+ * where both contra-variant and co-variant indices are shown as lower indices.
+ */
+template <typename DataType, typename Symm, typename IndexList>
+Scalar<DataType> pointwise_l2_norm(
+    const Tensor<DataType, Symm, IndexList>& tensor) noexcept {
+  return Scalar<DataType>{
+      sqrt(get(L2Norm_detail::pointwise_l2_norm_square(tensor)))};
+}
+
+// @{
+/*!
+ * \ingroup TensorGroup
+ * \brief Compute Euclidean \f$L^2\f$-norm of arbitrary Tensors reduced over an
+ * element.
+ *
+ * \details
+ * Computes the RMS value of the point-wise Frobenius norm of a given Tensor
+ * with arbitrary rank over all grid points in an element. If the Tensor \f$A\f$
+ * has rank \f$n\f$ and dimensionality \f$D\f$, and the element has \f$N\f$
```

Comment:
this is a fair point.  I can use `N_p` for number of points or just use `N` as the summation's upper limit instead of `N-1` and state clearly that `... and the element (of order \f$N\f$) has \f$N+1\f$ points...`. Let me know which you think would be more coherent with the rest of the code. I mildly prefer the latter since it refrains from introducing yet another variable `N_p` when `N+1` should suffice.

Reply:
I agree, using `N+1` seems better, but I'm okay with either :)

Reply:
changed